Leaky and Albedo Boundary Conditions Implementation

nutcasev15 opened the pull request:
<!--
If you are a first-time contributor to OpenMC, please have a look at our
contributing guidelines:
https://github.com/openmc-dev/openmc/blob/develop/CONTRIBUTING.md
-->

# Description

Recently, there is an interest in the nuclear community to model microreactors for terrestrial and space applications. Often, these microreactors deploy neutron reflectors to reduce neutron leakage. This pull request implements boundary conditions to model that behaviour without physically modelling the reflector. Further, it also implements particle weight modification when it encounters a boundary, similar to the albedo feature in SERPENT. Regression tests run with ENDF VII.1 are included for both features.

To ensure that the albedo feature functions similarly to SERPENT, eigenvalue calculations for the regression test geometry were also run in SERPENT 2.1.31 for various boundary albedos using the ENDF VII.1 data library:

| Reflective Boundary Albedo |    SERPENT 2.1.31   |        OpenMC       |
|:--------------------------:|:-------------------:|:-------------------:|
|              0             | 0.17988 +/- 0.00034 | 0.18024 +/- 0.00017 |
|            0.952        | 1.58887 +/- 0.00031 | 1.58948 +/- 0.00024 |
|              1             | 2.27663 +/- 0.00035 | 2.27696 +/- 0.00028 |

The simulations were run for 100 batches with 10 inactive batches at 10000 particles per batch. The results show excellent agreement between SERPENT and OpenMC. As with SERPENT, boundary albedos are also implemented for periodic planar surfaces with similar agreement in the calculated neutron multiplication factor. I hope these features enable OpenMC to be suited for analysis of systems where neutron reflection or leakage is a critical factor.

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
<!--
While tests will automatically be checked by CI, it is good practice to
ensure that they pass locally first. See instructions here:
https://docs.openmc.org/en/latest/devguide/tests.html
-->

gridley commented on include/openmc/surface.h:
```diff
@@ -86,6 +86,8 @@ class Surface {
   int id_;                                          //!< Unique ID
   std::string name_;                                //!< User-defined name
   std::shared_ptr<BoundaryCondition> bc_ {nullptr}; //!< Boundary condition
+  std::double_t bc_alb_;                            //!< Boundary albedo
```
Don't store this on the surface; you should instead store it on the boundary condition object.

nutcasev15 replied:
I've updated the pull request. Moved it out of the `Surface` class

gridley reviewed (changes requested):
I do not believe the `LeakyBC` should be added. Adding an `AlbedoBC` would be reasonable, however. I think you should just store the albedo data in a class that inherits from `ReflectiveBC`, adding the albedo data, calling the original particle handling code, then multiplying the weight by the albedo.

`LeakyBC` should have the same mean behavior as an albedo approach but with higher variance and a little more computational efficiency. Unless there are applications that seriously benefit from this that I'm unaware of, we should definitely not add this extra code as the added simulation capability over a plain albedo setup is zero.

Lastly, why not just actually model the reflector? You're using an MC code on a leaky geometry--running the particles through the reflector is not going to add too much to the run time.

nutcasev15 commented:
> I do not believe the `LeakyBC` should be added. Adding an `AlbedoBC` would be reasonable, however. I think you should just store the albedo data in a class that inherits from `ReflectiveBC`, adding the albedo data, calling the original particle handling code, then multiplying the weight by the albedo.
> 
> `LeakyBC` should have the same mean behavior as an albedo approach but with higher variance and a little more computational efficiency. Unless there are applications that seriously benefit from this that I'm unaware of, we should definitely not add this extra code as the added simulation capability over a plain albedo setup is zero.
> 
> Lastly, why not just actually model the reflector? You're using an MC code on a leaky geometry--running the particles through the reflector is not going to add too much to the run time; the reflector will also change the neutron energy spectrum on top of having an albedo effect, so albedo alone is probably not a great modeling technique for small cores.

I can make the requested changes but there are two issues:
1. If `AlbedoBC` inherits from `ReflectiveBC`, we lose albedo capability on periodic surfaces. Another class would be required to implement that. Given that `PeriodicBC` is already complex, I chose to modify the `Surface` class instead. Ultimately, two new BC classes can be created as you recommend.
2. BC objects are shared between surfaces in OpenMC. If `AlbedoBC` is implemented, it would constrain the user to only one albedo value on any surface using that BC. While that is serviceable, usually albedo and leakage are different for different directions which would reduce the usefulness of the feature. It would also make the implementation different from SERPENT which was my original reference code.

As for `LeakyBC`, I fully agree with the fact that computationally there is no difference. This feature is not intended for high fidelity simulation or performance optimization. I wrote this feature because I wanted to quantitatively study neutron leakage rates for reactors to study the effectiveness of reflector materials and geometries. As a user, I can explicitly specify leakage rates that have physical meaning. Explicit leakage rates have the effect of letting me study the reactor geometry's contribution to leakage and the reflector's contribution to leakage independently. Eigenvalue simulations with a physical reflector and a `VacuumBC` struggle to resolve those two phenomena. Further, while there exists an equivalent albedo value to obtain the same results, as a user I have no method of obtaining the required albedo value from physical analysis.

It would be helpful if someone who extensively uses OpenMC for fixed source simulations could also pitch in, so that we also have another perspective as we have only talked about reactor simulation so far. I suspect there are possible applications for `LeakyBC` and 3D albedos there too.

gridley commented:
Why would you want an albedo on a periodic surface? I can't imagine any situation where that would be a physically useful model. Periodicity is only used for symmetries (AFAIK), so an albedo effect cannot come into play. Maybe I'm not seeing the whole picture...

Regarding the possibility of differing albedo values, there's nothing stopping us from constructing another BC object for different values of the albedo. They're all held by smart pointers so it should be agnostic to having either one shared or many unique BC objects. So that should resolve those issues, maybe?

As for studying the reactor geometry vs. reflector contribution to leakage independently, IMO wouldn't it be better to tally partial currents on the surface that separates the core and reflector? That's not to say albedo BCs are not useful; I have hoped to add them for a while! Well, I don't think we can tally partial currents separately now. It would be nice to add that.

nutcasev15 commented:
> Why would you want an albedo on a periodic surface? I can't imagine any situation where that would be a physically useful model. Periodicity is only used for symmetries (AFAIK), so an albedo effect cannot come into play. Maybe I'm not seeing the whole picture...
> 
> Regarding the possibility of differing albedo values, there's nothing stopping us from constructing another BC object for different values of the albedo. They're all held by smart pointers so it should be agnostic to having either one shared or many unique BC objects. So that should resolve those issues, maybe?
> 
> As for studying the reactor geometry vs. reflector contribution to leakage independently, IMO wouldn't it be better to tally partial currents on the surface that separates the core and reflector? That's not to say albedo BCs are not useful; I have hoped to add them for a while! Well, I don't think we can tally partial currents separately now. It would be nice to add that.

Your comments pushed me to look and I found this: https://www.sciencedirect.com/science/article/pii/S0306454915004934
That study is a parametric study of neutron leakage with SERPENT. While modelling hexagonal fuel elements, a periodic boundary with an alebedo comes into play. I think its best to include that for completeness, even if its a rare use case.

I was confused with the shared pointer C++ but yes, there will be no conflicts. I'll modify the pull request to have `AlbedoBC` , `PeriodicAlbedoBC` and `LeakyBC` with BC data in their separate classes. Storage and retrieval of BC data will still be tied to the `Surface` object though, if that works.

gridley commented:
I really do not think we should add more data to the `Surface` class. There are models out there with loads of surfaces in them, and that would make their memory consumption potentially grow by a large amount.

nutcasev15 commented:
> Totally agree, if it's in the literature and isn't too hard let's add it!
> 
> On the partial current: you can get the partial current in both directions to explore the incoming reflected neutrons vs. outgoing leakage.
> 
> Also, you should not need to duplicate code between the AlbedoReflective and AlbedoPeriodic case. This is a nice one to solve with a decorator template! Then we should have zero replicated code. Check this out:
> 
> ```
> template <typename BaseBC>
> class AlbedoBC : public BaseBC {
> public:
>   AlbedoBC() = delete; // prevent creation without albedo set
>   AlbedoBC(double albedo) : albedo_(albedo) {}
> 
>   virtual void handle_particle(Particle& p, const Surface& surf) override {
>     BaseBC::handle_particle(p, surf);
>     p.wgt() *= albedo_;
>   }
> private:
>   double albedo_;
> };
> ```
> 
> Then just do:
> 
> ```
> typedef AlbedoBC<PeriodicBC> AlbedoPeriodicBC;
> typedef AlbedoBC<ReflectiveBC> AlbedoReflectiveBC;
> ```
> 
> Etc. And you could even tack it onto the white BC if you want to play with reflector modeling that loses directional information. I could imagine that might actually be a nice model of a reflector. 🤔

Thank you @gridley ! This is a fantastically clean way to do it. I'm not as advanced with C++ but I thought this method would have issues with reading and writing the albedo values from the XMLs. I therefore modified the BoundaryCondition class instead. The updated code moves everything into BoundaryCondition or its children. Since they are under std::shared_ptr, I think that should address the memory issue. Are you happy with the code as it stands now?

gridley reviewed (changes requested):
Definitely should remove `leaky` as a BC type, and definitely should use the decorator approach for the BCs. Adding an albedo value to every single BC is just really not clean. The fact that we have to give an albedo to a vacuum BC is just nasty--it's like if we were writing the code in fortran or something.

nutcasev15 commented:
Thank you for the detailed comments! I will redo this with the decorator approach. There will still be a `get_albedo` virtual in `BoundaryCondition` because I need that to make the `write_hdf5` function in `Surface` to work. `Surface` assumes a shared_ptr of `BoundaryCondition` which is fine until a class derived from `BoundaryCondition` implements new functions or variables. `bc_` in `write_hdf5` cannot access the `albedo` variable otherwise.

Currently, I need `LeakyBC` for my work. In general, I think having the ability to set boundary leakage is very useful for reactor design and scoping simulations. If `LeakyBC` is removed, how can the user set a particular leakage rate using only the albedo BCs?

nutcasev15 commented on src/surface.cpp:
```diff
@@ -1156,9 +1196,10 @@ void read_surfaces(pugi::xml_node node)
+  // periodic surfaces can only have one albedo value.
   model::surfaces.reserve(n_surfaces);
-  std::set<std::pair<int, int>> periodic_pairs;
+  std::set<std::pair<std::pair<int, int>, std::double_t>> periodic_pairs;
```
I did this to ensure I don't mess up the albedos between pairs of periodic surfaces because there is rarely only one pair. How would you suggest I rewrite this?

gridley replied:
I would suggest adding an `std::map` that maps surface IDs to albedo values. Then when the PeriodicBC objects are created, try to lookup albedo values from the map. Does that make sense?

gridley commented:
The solution to that is not put albedo on every single `BoundaryCondition`, but rather to add a virtual method to `BoundaryCondition` called `to_hdf5_inner`, similar to how we differentiate writing the different surfaces to HDF5. That would be very nice and clean. Then you'd just do:

```
if (bc_) bc_->to_hdf5_inner(group);
```

What do you think?

As for LeakyBC, let's get some input from @paulromano or @pshriwise or someone else. IMO you should probably just keep it in your own branch of the code, because (after 7 years of doing reactor scoping studies) that does not seem like a useful feature to me when albedo options are already present. The only difference would be that the leakage output has leaked neutrons for the LeakyBC but not the albedo BCs.

nutcasev15 commented:
> The solution to that is not put albedo on every single `BoundaryCondition`, but rather to add a virtual method to `BoundaryCondition` called `to_hdf5_inner`, similar to how we differentiate writing the different surfaces to HDF5. That would be very nice and clean. Then you'd just do:
> 
> ```
> if (bc_) bc_->to_hdf5_inner(group);
> ```
> 
> What do you think?
> 
> As for LeakyBC, let's get some input from @paulromano or @pshriwise or someone else. IMO you should probably just keep it in your own branch of the code, because that does not seem like a useful feature to me when albedo options are already present. The only difference would be that the leakage output has leaked neutrons for the LeakyBC but not the albedo BCs.

That is a very good solution that didn't occur to me. I thought having a virtual `get_albedo()` and `get_leakage()` method. That way the data can be queried out of the BCs by `TallyFilter` or `Surface` classes, should they need to in the future. Should I leave it as it is or do it the way you suggest? I think having a virtual getter function is more flexible but the `to_hdf5_inner` solves the immediate problem and is cleaner.

If after input from the others, the verdict is in favour of removing `LeakyBC`, I'll definitely do that. As it stands, with your code, I have compressed things down quite a bit and made a decorator template which gives `LeakyReflectiveBC` and `LeakyWhiteBC`. The C++ side is only 20 extra lines in the header over the Albedo BC stuff. As you said, if its easy to implement, we should do it.

I know that killing particles is equivalent to reducing all the particle's importance but I'm not aware of any method to find the albedo for a given geometry if I require say 50 % leakage. Further, I'm not aware of any code that allows setting leakages and my thesis indicates that it is useful for microreactor design, a field that is gathering steam and will be important in the future.

I was curious about the partial flux tally and it seems to me that maybe a new implementation is not needed. A tally with `SurfaceFilter` and `MuFilter` would have the same behaviour, no?

Also, I'd love for you to explain some more on how I can avoid the `double` in the `periodic_pairs` when dealing with `PeriodicBC` and not mess up albedos if there are multiple pairs?

paulromano commented:
Thanks for the ping and thanks @nutcasev15 for the contribution. I'm mostly off this week but would like to take a look at this -- hope to get you some feedback early next week!

gridley commented:
Ideally `TallyFilter` and `Surface` won't need to know about specific albedo values for encapsulation reasons.

I think that setting the leak fraction on `LeakyBC` will not necessarily guarantee a certain value of the leakage. For a very small problem it approximately will, but imagine the case of a perfect absorber set before the `LeakyBC`. Clearly in that case, the leakage is zero regardless of the leak fraction you set on the BC. In the general case, finding a given leakage fraction would be a stochastic root finding loop over a parameter like reactor size or an albedo value. Could you share a link to your thesis where you found manually setting this to be useful? Would very much be interested as my preconceptions around this could be a bit flawed, so I apologize for my hesitance in accepting its utility.

I'm pretty sure that `MuFilter` bins the scattering direction cosine, so it wouldn't help with partial currents. It would definitely be best suited as its own score. It's something for us to be thinking about...

As for explaining removing the double in the pairs, why not save the albedo value to an `std::map` using the surface ID as a key and the albedo as the value? Then you can just access that data when constructing the BC in the place where the loop over ID pairs happens. If there is an inconsistency in the albedo value for a periodic problem, you'd just need to throw an error. You could have two different weight multipliers on each surface, which seems like it would be a little strange, but perhaps not an unreasonable model of certain depleted problems. I can provide a code example or make a PR to your PR if that still isn't clear.

nutcasev15 commented:
> Ideally `TallyFilter` and `Surface` won't need to know about specific albedo values for sake of encapsulation.
> 
> I think that setting the leak fraction on `LeakyBC` will not necessarily guarantee a certain value of the leakage. For a very small problem it approximately will, but imagine the case of a perfect absorber set before the `LeakyBC`. Clearly in that case, the leakage is zero regardless of the leak fraction you set on the BC. In the general case, finding a given leakage fraction would be a stochastic root finding loop over a parameter like reactor size or an albedo value. Could you share a link to your thesis where you found manually setting this to be useful? Would very much be interested as my preconceptions around this could be a bit flawed, so I apologize for my hesitance in accepting its utility.
> 
> I'm pretty sure that `MuFilter` bins the scattering direction cosine, so it wouldn't help with partial currents. It would definitely be best suited as its own score. It's something for us to be thinking about...
> 
> As for explaining removing the double in the pairs, why not save the albedo value to an `std::map` using the surface ID as a key and the albedo as the value? Then you can just access that data when constructing the BC in the place where the loop over ID pairs happens. If there is an inconsistency in the albedo value for a periodic problem, you'd just need to throw an error. You could have two different weight multipliers on each surface, which seems like it would be a little strange, but perhaps not an unreasonable model of certain depleted problems. I can provide a code example or make a PR to your PR if that still isn't clear.

I'm very grateful for your guidance so far. I've redone the pull request using your code and direction. Leakage is still included, for now.

My thesis is available here: https://github.com/nutcasev15/MSc-Thesis
Chapter 6 is relevant here. My thesis systematically evaluates moderator materials but the next step is reflector materials. The parametric model `Core` allows me to play with reflector thicknesses and materials.  As you said, the reflector alters the neutron spectrum. However, I can imagine scenarios (Optimum moderator to fuel ratio reduces leakage to nearly 0; Thermal spectrum reactor will not be strongly effected; BeO moderation effect is minimal) where it wouldn't matter. Leakage specification comes into the picture when I want to do a reactor geometry sweep. My thesis assumes a constant R/H ratio of the reactor based on an ideal value in literature. The issue is that changing the reactor radius while keeping the reflector radius constant does not imply the same leakage. So I need to set a constant value there so that I can compare results between two reactor sizes.

My worry is that if I set a specific albedo, it not equal to modelling leakage in the physical sense. I will be able to compare between geometries but my results cannot be used to infer general design results. My 0.5 albedo will not have the same physical behaviour as someone else's 0.5 albedo with their design. However, if I set 50 % leakage and say XYZ behaves like ABC, then the reader can expect ABC behaviour with their design if they use 50 % leakage. This is my understanding but I'm aware that you have a lot more experience than me, so it might turn out I'm worried for no reason.

Besides that, I've realised that I'll need the partial current measurement score. OpenMC documentation actually talks about this and I think its already implemented: https://docs.openmc.org/en/stable/usersguide/tallies.html

![image](https://github.com/openmc-dev/openmc/assets/20016119/4f459897-5c22-44a7-850d-b5e4a7a57314)

HunterBelanger commented:
I personally find it very odd that the `AlbedoBC` does not contribute to the leakage score, and if I understand correctly, this is actually the fundamental difference and argument for having both `LeakyBC` and `AlbedoBC`, correct ?

The two ways to handle a known leakage rate BC of L (with L in [0,1]) would be to "leak" the particle with probability `L` and reflect with the complementary probability. This is of course the current `LeakyBC` implementation.

The other way of handling such a BC is contributing L*w to the leakage score, multiplying the weight of the particle by (1-L), and then reflecting. Currently, `AlbedoBC` does the second portion of this, but it does not contribute to the leakage score. I do not think it makes any sense to simply reduce the weight of particles at an albedo BC and effectively state that this removed weight has disappeared and does not contribute to the leakage score at all.  I think `AlbedoBC` should be modified to also contribute to leakage scores, and once that is done, `LeakyBC` and `AlbedoBC` will be completely equivalent from a physics standpoint.

One difference that would still remain between `LeakyBC` and `AlbedoBC` would then be their use in analog transport. If you want to be an analog purist, then `LeakyBC` makes more sense when using analog transport. I believe that `AlbedoBC` should work just fine with the current analog transport algorithms, it is just that particles will end up not having a weight of 1, which is usually expected in analog transport. I am not sure I have a strong opinion on this subject, I just wanted to point it out.

Another difference is that `AlbedoBC` is actually more powerful (and therefore dangerous). In `LeakyBC`, you could only have 0 < L < 1, however, with `AlbedoBC`, you can have L take any value. If L < 0, your leakage score can be negative. IMO this is not necessarily problematic, and simply means you do not actually have net leakage, but a net gain of neutrons from the outside. More problematic however is when L > 1. Here, the weight of the particles can become negative. While I personally do transport with negative weights all the time in my own MC code, I have specialized weight cancellation algorithms to allow this, that OpenMC does not have. You can read more about this [here](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.104.015306) if you want, but without getting into details, allowing negative particles in power iteration will lead to an explosion in the number of particles, and the program will eventually crash. As such, a hard error should definitely be raised (in both Python and C++) if a user gives L > 1 with an albedo BC (or as currently written, an albedo factor < 0).

HunterBelanger commented on docs/source/usersguide/geometry.rst:
```diff
@@ -196,6 +214,35 @@ lies in the first quadrant of the Cartesian grid. If the geometry instead lies
+can be used to modify the importance of particles that encounter the boundary.
+The albedo value specifies the ratio between the particle's importance after
+interaction with the boundary to its initial importance. Currently, OpenMC
+only accepts albedos between 0 and 1. The following example creates a reflective
```
An error should definitely be raised if the albedo is < 0, in both Python and C++. My comment below explains why.

gridley replied:
Makes sense!

nutcasev15 commented:
Thank you very much @gridley and @HunterBelanger for your comments and input. I've updated the `AlbedoBC` implementation to properly tally leakages and particle weights. As @HunterBelanger pointed out, this makes `AlbedoBC` physically equivalent to `LeakyBC` and therefore, I've removed it. The appropriate warnings and hard errors are also included in the Python API and in the C++ `Surface` class.

Finally, here are the regression test results using ENDF VII.1:
| Reflective Boundary Albedo |    SERPENT 2.1.31   |        OpenMC       |    OpenMC Leakage    |
|:--------------------------:|:-------------------:|:-------------------:|:--------------------:|
|              0             | 0.17988 +/- 0.00034 | 0.18024 +/- 0.00017 |  0.92688 +/- 0.00031 |
|            0.952           | 1.58887 +/- 0.00031 | 1.58948 +/- 0.00024 |  0.31456 +/- 0.00023 |
|              1             | 2.27663 +/- 0.00035 | 2.27696 +/- 0.00028 |  0.00000 +/- 0.00000 |
|            1.05            |          -          | 3.92576 +/- 0.00127 | -0.77350 +/- 0.00135 |

Again, the agreement in excellent. There are 2 additional things to note; leakage at 0.952 albedo is not the same as leakage at 0 albedo multiplied with 0.952. I believe that's expected but it is interesting and deserving of more research. It suggests that, aside from geometry and boundary albedo, there are other factors which determine overall leakage. Lastly, an albedo above unity causes an influx of neutrons, as expected. OpenMC is touchy with this however, with fission bank overflow warnings all over the place if I use an albedo greater than 1.05.

nutcasev15 commented:
@gridley and @paulromano Please let me know if any further changes are required or if this is ready for a merge.

HunterBelanger commented:
Hi @nutcasev15, I haven't been able to give this a thorough review (@gridley has already given you a lot of great feedback), but I am somewhat confused about the case you mention with the albedo = 1.05, and getting lots of fission bank overflow warnings. I would expect this for the first few generations for a system with such a large keff, but since the number of generated fission particles is normalized by keff, once you have converged after a few generations, I would expect this normalization to prevent these overflow warnings later on in the simulation. After the simulation has converged, do these overflow warnings disappear, or do they continue ? If they keep appearing, I am not sure I immediately understand why this is the case.

nutcasev15 commented:
> Hi @nutcasev15, I haven't been able to give this a thorough review (@gridley has already given you a lot of great feedback), but I am somewhat confused about the case you mention with the albedo = 1.05, and getting lots of fission bank overflow warnings. I would expect this for the first few generations for a system with such a large keff, but since the number of generated fission particles is normalized by keff, once you have converged after a few generations, I would expect this normalization to prevent these overflow warnings later on in the simulation. After the simulation has converged, do these overflow warnings disappear, or do they continue ? If they keep appearing, I am not sure I immediately understand why this is the case.

Once the inactive batches are done, those warnings disappear. If the albedo is set to more extreme values, say 1.2, the k value never converges and I get warnings for all the batches I run.

gridley commented:
Ah, yes, it makes sense that a fission source would diverge in that case. Because k is only adjusting the amount of fission neutrons that get produced from generation to generation, it doesn't have any influence over how weight changes when crossing a boundary. We calculate the number of fission sites created at a collision site in a way that depends on the weight.

If the albedo too far above one, the weights can exponentially increase without regard to the k eigenvalue. You could observe this in an arbitrarily small volume with reflective BCs and albedo >1. Since the boundary is crossed many times the weight grows very much, and many fission sites are created. The calculation then increases k to compensate for the large number of progeny in the next generation.

If the system is such that with zero fission it remains "supercritical" due to multiplication at the albedo BC, it will diverge.

So that would explain the issue, I think.

gridley reviewed (changes requested):
This is starting to look very nice to me; thank you for the contribution! I do have a few more minor revisions to suggest before merging.

gridley commented on include/openmc/boundary_condition.h:
```diff
@@ -106,5 +113,49 @@ class RotationalPeriodicBC : public PeriodicBC {
+//! Compatible with ReflectiveBC, WhiteBC and both types of PeriodicBC.
+//==============================================================================
+
+template<typename BaseBC>
+class AlbedoBC : public BaseBC {
+public:
+  AlbedoBC(double albedo) : bc_alb_(albedo) {};
+  AlbedoBC(int i_surf, int j_surf, double albedo)
+    : BaseBC(i_surf, j_surf), bc_alb_(albedo) {};
+  // Default to not modifying particle weights
+  AlbedoBC() : bc_alb_(1.0) {};
+
+  void handle_particle(Particle& p, const Surface& surf) const override
+  {
+    // Save incident particle's weight
+    double initial_wgt = p.wgt();
+    // Treat the lost weight fraction as leakage, similar to VacuumBC.
+    // This ensures the lost weight is tallied properly.
+    p.wgt() *= (1.0 - bc_alb_);
+    p.cross_vacuum_bc(surf);
+    // Reassign the remaining weight and pass particle to BaseBC
+    p.wgt() = initial_wgt * bc_alb_;
+    BaseBC::handle_particle(p, surf);
+  };
```
In what sense is it overkill? I suggested this so as to not have irrelevant data on a class like VacuumBC, for example.

paulromano replied:
Overkill in the sense that it deepens the class hierarchy when the existing classes could make do, and it's more to track mentally for someone reading the code. Vacuum BC is the only case where it doesn't make sense; to handle that we can have a default albedo of 1.0 and only invoke the albedo handling if the value is not 1.0.

gridley replied:
Alright, sounds reasonable to me!

paulromano replied:
@nutcasev15 does that sound acceptable to you?

nutcasev15 replied:
The first revision of this PR was very similar to what you have suggested. However, we determined that the template way was more modular, reduced the fragmentation of code changes across files and used slightly less memory and ran slightly faster.

Ultimately, I think the memory and speed criteria are the most important and I've done a comparison:

**Current `<template>` Implementation:**
No Albedo in XML (Legacy Models):
- Calculation Rate (inactive)       = 76830.7 particles/second
- Calculation Rate (active)         = 78405.9 particles/second

Albedo is 1.0 in XML:
- Calculation Rate (inactive)       = 77180.6 particles/second
- Calculation Rate (active)         = 80555.1 particles/second

**Suggested `BoundaryCondition` Modification Implementation:**
No Albedo in XML (Legacy Models):
- Calculation Rate (inactive)       = 76682.5 particles/second
- Calculation Rate (active)         = 77538.1 particles/second

Albedo is 1.0 in XML:

- Calculation Rate (inactive)       = 76869.9 particles/second
- Calculation Rate (active)         = 78743.8 particles/second

I think we should err on the side of performance because it can have knock on effects on depletion runs and other repetitive eigenvalue calculations. Further, the current implementation ensures that no legacy models will break while making changes to `BoundaryCondition` means users that don't use albedos would still have a slight performance penalty. I vote to go ahead with how things are currently.

The diff of the changes for the `BoundaryCondition` Modification implementation is below:
```
diff --git a/include/openmc/boundary_condition.h b/include/openmc/boundary_condition.h
index fb35b87a8..668f14f0a 100644
--- a/include/openmc/boundary_condition.h
+++ b/include/openmc/boundary_condition.h
@@ -4,7 +4,6 @@
 #include "openmc/hdf5_interface.h"
 #include "openmc/particle.h"
 #include "openmc/position.h"
-#include "openmc/random_lcg.h"
 #include <fmt/core.h>
 
 namespace openmc {
@@ -24,13 +23,39 @@ public:
   //!   to directly modify anything about the particle, but it will do so
   //!   indirectly by calling the particle's appropriate cross_*_bc function.
   //! \param surf The specific surface on the boundary the particle struck.
-  virtual void handle_particle(Particle& p, const Surface& surf) const = 0;
+  virtual void handle_particle(Particle& p, const Surface& surf) const {
+
+    if (!has_alebedo_)
+      return;
+
+    // Save incident particle's weight
+    double initial_wgt = p.wgt();
+    // Treat the lost weight fraction as leakage, similar to VacuumBC.
+    // This ensures the lost weight is tallied properly.
+    p.wgt() *= (1.0 - albedo_);
+    p.cross_vacuum_bc(surf);
+    // Reassign the remaining weight and pass particle to BaseBC
+    p.wgt() = initial_wgt * albedo_;
+  };
 
   //! Return a string classification of this BC.
   virtual std::string type() const = 0;
 
   //! Write BC data such as albedo to hdf5.
-  virtual void to_hdf5(hid_t surf_group) const {};
+  void to_hdf5(hid_t surf_group) const  {
+    if (std::abs(albedo_ - 1.0) <= FP_PRECISION) {
+      write_string(surf_group, "albedo", fmt::format("{}", albedo_), false);
+    }
+  };
+
+  virtual void set_albedo(double albedo) {
+    has_alebedo_ = true;
+    albedo_ = albedo;
+  }
+
+private:
+double albedo_ = 1.0;
+bool has_alebedo_ = false;
 };
 
 //==============================================================================
@@ -113,49 +138,5 @@ protected:
   double angle_;
 };
 
-//==============================================================================
-//! A BC template that modifies the weight of incident particles.
-//
-//! Compatible with ReflectiveBC, WhiteBC and both types of PeriodicBC.
-//==============================================================================
-
-template<typename BaseBC>
-class AlbedoBC : public BaseBC {
-public:
-  AlbedoBC(double albedo) : bc_alb_(albedo) {};
-  AlbedoBC(int i_surf, int j_surf, double albedo)
-    : BaseBC(i_surf, j_surf), bc_alb_(albedo) {};
-  // Default to not modifying particle weights
-  AlbedoBC() : bc_alb_(1.0) {};
-
-  void handle_particle(Particle& p, const Surface& surf) const override
-  {
-    // Save incident particle's weight
-    double initial_wgt = p.wgt();
-    // Treat the lost weight fraction as leakage, similar to VacuumBC.
-    // This ensures the lost weight is tallied properly.
-    p.wgt() *= (1.0 - bc_alb_);
-    p.cross_vacuum_bc(surf);
-    // Reassign the remaining weight and pass particle to BaseBC
-    p.wgt() = initial_wgt * bc_alb_;
-    BaseBC::handle_particle(p, surf);
-  };
-
-  void to_hdf5(hid_t surf_group) const override
-  {
-    if (std::abs(bc_alb_ - 1.0) <= FP_PRECISION) {
-      write_string(surf_group, "albedo", fmt::format("{}", bc_alb_), false);
-    }
-  };
-
-private:
-  double bc_alb_;
-};
-
-typedef AlbedoBC<ReflectiveBC> ReflectiveAlbedoBC;
-typedef AlbedoBC<WhiteBC> WhiteAlbedoBC;
-typedef AlbedoBC<RotationalPeriodicBC> RotationalPeriodicAlbedoBC;
-typedef AlbedoBC<TranslationalPeriodicBC> TranslationalPeriodicAlbedoBC;
-
 } // namespace openmc
 #endif // OPENMC_BOUNDARY_CONDITION_H
diff --git a/src/boundary_condition.cpp b/src/boundary_condition.cpp
index 5b842399e..fa55f4e79 100644
--- a/src/boundary_condition.cpp
+++ b/src/boundary_condition.cpp
@@ -28,6 +28,7 @@ void ReflectiveBC::handle_particle(Particle& p, const Surface& surf) const
   Direction u = surf.reflect(p.r(), p.u(), &p);
   u /= u.norm();
 
+  BoundaryCondition::handle_particle(p, surf);
   p.cross_reflective_bc(surf, u);
 }
 
@@ -40,6 +41,7 @@ void WhiteBC::handle_particle(Particle& p, const Surface& surf) const
   Direction u = surf.diffuse_reflect(p.r(), p.u(), p.current_seed());
   u /= u.norm();
 
+  BoundaryCondition::handle_particle(p, surf);
   p.cross_reflective_bc(surf, u);
 }
 
@@ -130,6 +132,8 @@ void TranslationalPeriodicBC::handle_particle(
       "hitting a surface, but that surface is not recognized by the BC.");
   }
 
+  BoundaryCondition::handle_particle(p, surf);
+
   // Pass the new location and surface to the particle.
   p.cross_periodic_bc(surf, new_r, p.u(), new_surface);
 }
@@ -264,6 +268,8 @@ void RotationalPeriodicBC::handle_particle(
   Direction new_u = {
     cos_theta * u.x - sin_theta * u.y, sin_theta * u.x + cos_theta * u.y, u.z};
 
+  BoundaryCondition::handle_particle(p, surf);
+
   // Pass the new location, direction, and surface to the particle.
   p.cross_periodic_bc(surf, new_r, new_u, new_surface);
 }
diff --git a/src/surface.cpp b/src/surface.cpp
index 8adab6f6a..a0295377e 100644
--- a/src/surface.cpp
+++ b/src/surface.cpp
@@ -77,8 +77,25 @@ Surface::Surface(pugi::xml_node surf_node)
   if (check_for_node(surf_node, "boundary")) {
     std::string surf_bc = get_node_value(surf_node, "boundary", true, true);
 
+    if (surf_bc == "transmission" || surf_bc == "transmit" || surf_bc.empty()) {
+      // Leave the bc_ a nullptr
+    } else if (surf_bc == "vacuum") {
+      bc_ = std::make_shared<VacuumBC>();
+    } else if (surf_bc == "reflective" || surf_bc == "reflect" ||
+               surf_bc == "reflecting") {
+      bc_ = std::make_shared<ReflectiveBC>();
+    } else if (surf_bc == "white") {
+      bc_ = std::make_shared<WhiteBC>();
+    } else if (surf_bc == "periodic") {
+      // Periodic BCs are handled separately
+    } else {
+      fatal_error(fmt::format("Unknown boundary condition \"{}\" specified "
+                              "on surface {}",
+        surf_bc, id_));
+    }
+
     double surf_alb;
-    if (check_for_node(surf_node, "albedo")) {
+    if (check_for_node(surf_node, "albedo") && bc_ != nullptr) {
       surf_alb = std::stod(get_node_value(surf_node, "albedo"));
 
       if (surf_alb < 0.0)
@@ -91,29 +108,8 @@ Surface::Surface(pugi::xml_node surf_node)
                             "Albedos greater than 1 may cause "
                             "unphysical behaviour.",
           id_, surf_alb));
-    }
 
-    if (surf_bc == "transmission" || surf_bc == "transmit" || surf_bc.empty()) {
-      // Leave the bc_ a nullptr
-    } else if (surf_bc == "vacuum") {
-      bc_ = std::make_shared<VacuumBC>();
-    } else if (surf_bc == "reflective" || surf_bc == "reflect" ||
-               surf_bc == "reflecting") {
-      if (check_for_node(surf_node, "albedo"))
-        bc_ = std::make_shared<ReflectiveAlbedoBC>(surf_alb);
-      else
-        bc_ = std::make_shared<ReflectiveBC>();
-    } else if (surf_bc == "white") {
-      if (check_for_node(surf_node, "albedo"))
-        bc_ = std::make_shared<WhiteAlbedoBC>(surf_alb);
-      else
-        bc_ = std::make_shared<WhiteBC>();
-    } else if (surf_bc == "periodic") {
-      // Periodic BCs are handled separately
-    } else {
-      fatal_error(fmt::format("Unknown boundary condition \"{}\" specified "
-                              "on surface {}",
-        surf_bc, id_));
+      bc_->set_albedo(surf_alb);
     }
   }
 }
@@ -1334,31 +1330,19 @@ void read_surfaces(pugi::xml_node node)
     // condition.  Otherwise, it is a rotational periodic BC. If albedo data
     // is present, instantiate appropriate BC type.
     if (std::abs(1.0 - dot_prod) < FP_PRECISION) {
-      if (albedo_map.count(surf1.id_)) {
-        surf1.bc_ = std::make_shared<TranslationalPeriodicAlbedoBC>(
-          i_surf, j_surf, albedo_map[surf1.id_]);
-      } else {
-        surf1.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
-      }
-      if (albedo_map.count(surf2.id_)) {
-        surf2.bc_ = std::make_shared<TranslationalPeriodicAlbedoBC>(
-          i_surf, j_surf, albedo_map[surf2.id_]);
-      } else {
-        surf2.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
-      }
+      surf1.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
+      surf2.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
     } else {
-      if (albedo_map.count(surf1.id_)) {
-        surf1.bc_ = std::make_shared<RotationalPeriodicAlbedoBC>(
-          i_surf, j_surf, albedo_map[surf1.id_]);
-      } else {
-        surf1.bc_ = std::make_shared<RotationalPeriodicBC>(i_surf, j_surf);
-      }
-      if (albedo_map.count(surf2.id_)) {
-        surf2.bc_ = std::make_shared<RotationalPeriodicAlbedoBC>(
-          i_surf, j_surf, albedo_map[surf2.id_]);
-      } else {
-        surf2.bc_ = std::make_shared<RotationalPeriodicBC>(i_surf, j_surf);
-      }
+      surf1.bc_ = std::make_shared<RotationalPeriodicBC>(i_surf, j_surf);
+      surf2.bc_ = std::make_shared<RotationalPeriodicBC>(i_surf, j_surf);
+    }
+
+    if (albedo_map.count(surf1.id_)) {
+        surf1.bc_->set_albedo(albedo_map[surf1.id_]);
+    }
+
+    if (albedo_map.count(surf2.id_)) {
+        surf2.bc_->set_albedo(albedo_map[surf2.id_]);
     }
   }
 }
```

paulromano replied:
Thanks for the detailed analysis @nutcasev15. What is the test problem you are running here?

nutcasev15 replied:
Its the regression test that's included in the PR. I think it's a good representation of models where there is a large amount of interaction with the boundary conditions.

paulromano replied:
I tried out both implementations on my machine and I'm seeing very little difference in performance, at most 1–2%, although even then it's hard to get solid numbers because the normal machine jitter tends to be more than that. Keeping in mind that this is on a very contrived problem with particles bouncing back and forth off boundaries, for a practical problem we will see effectively 0 difference in performance.

Regarding memory, you can get away with storing just the `albedo_` member (whether a BC "has" an albedo can be checked by comparing to 1.0), which means that you have at most an extra 8 bytes per boundary condition. I've never seen a model with more than six explicitly specified BCs, so the memory cost is at most 48 bytes.

One worry I have with the template approach is its impact for a future GPU implementation. We have to explicitly instantiate each concrete class, which results in a bunch of extra code that has to be compiled that is most likely not used (on my system, the templated approach results in `libopenmc.so` being 200 kB larger). On CPU this is not a big deal, but on GPU you pay a price to have more code within a kernel because it affects register usage and hence occupancy.

So, overall I don't think the potentially small performance benefit on the test problem is worth the extra complexity and possible negative impact down the line for GPU porting purposes.

nutcasev15 replied:
I fully understand and agree with you that the performance gain minimal. PR updated @paulromano! I hope this is merge ready?

gridley replied:
Good point Paul on the GPU stuff, although I suspect the compiler would be able to eliminate the extra code if we went with a tagged union approach to replace polymorphism for this hypothetical nicely written GPU version. If the BC classes are just functors, it could inline the code in the switch/case statement that gets generated by a variant.

Regardless, I think putting the albedo on the BC makes a lot of sense. Previously it was on each surface... somehow I didn't think about this possibility. It can be tempting to jump to one of these fancy "design patterns" just to use some fun C++ features. 😄

paulromano commented on src/surface.cpp:
```diff
@@ -1303,10 +1330,18 @@ void read_surfaces(pugi::xml_node node)
     if (std::abs(1.0 - dot_prod) < FP_PRECISION) {
       surf1.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
-      surf2.bc_ = surf1.bc_;
+      surf2.bc_ = std::make_shared<TranslationalPeriodicBC>(i_surf, j_surf);
```
If periodic surfaces don't share a BoundaryCondition object, there is no need or them to be `shared_ptr` (could be `unique_ptr` instead). You don't necessarily need to make that change as part of this PR but just wanted to point this out while it's still in my mind so we can come back to it later.

nutcasev15 replied:
I'll fix this shortly

paulromano reviewed (changes requested):
Thanks for the updates! Other than the comments below, the only other request I have is that surface_composite.py will need to be updated to work with the new `albedo` argument wherever `boundary_type` is accepted.

nutcasev15 commented on include/openmc/boundary_condition.h:
```diff
@@ -16,14 +19,47 @@ class Surface;
+
+private:
+  double albedo_ = 1.0;
+  bool has_albedo_ = false;
```
I avoided that because comparing floating points like that might lead to undefined behaviour. Other places in the code do floating point comparison using abs(a - b) < minimum float value logic. I can do that but it will translate to more instructions. Given this check will return every time a particle is processed for any BC, I thought to just use an extra byte in the BC class and do it the way I've done it. Are there drawbacks to my approach that don't justify the optimisation I'm doing?

paulromano replied:
While it is true that floating point comparison is generally unreliable, it should be safe in this particular case because 1) the number 1.0 has an exact representation and 2) we are setting the default value using the same constant against which we are comparing. Of course, it wouldn't be reasonable to expect the result of some mathematical expression to be exactly 1.0 (e.g., `3*0.3 + 0.1`), but for just comparing against a constant I don't see how it could run into problems.

Now, that being said, a safer alternative that would avoid the `!=` comparison is to set the default value to `-1.0` as a "null" value, and then have `bool has_albedo() const { return albedo_ > 0.0; }`

nutcasev15 replied:
That is a sensible solution, will implement that instead.

nutcasev15 commented:
@paulromano I have updated the PR with the agreed upon changes

paulromano reviewed (approved):
Thanks @nutcasev15!

@gridley Do you have any further requests on this PR or do you approve?

gridley commented on include/openmc/boundary_condition.h:
```diff
@@ -16,14 +19,44 @@ class Surface;
   //!   indirectly by calling the particle's appropriate cross_*_bc function.
   //! \param surf The specific surface on the boundary the particle struck.
-  virtual void handle_particle(Particle& p, const Surface& surf) const = 0;
+  virtual void handle_particle(Particle& p, const Surface& surf) const
```
I do not like the approach of having a function named `handle_particle` only handle albedo. We should make this a non-virtual method called `handle_albedo`, then leave `handle_particle` as `=0`'d here.

nutcasev15 replied:
Fixed the PR @gridley ! That should do it.

paulromano commented:
@gridley Do you approve of the changes?

gridley commented:
This code looks fantastic now! Thanks @nutcasev15! 😄
